File: consumer-1/tests/registration.test.js
/**
 * CONSUMER REGISTRATION
 * ---------------------
 * Registers which API endpoints your service uses so breaking changes can be detected.
 *
 * Prerequisites:
 * - CVT server running
 *
 * Key concepts:
 * - Auto-registration: Capture interactions from mocks/adapters, then register
 * - Manual registration: Explicitly declare used endpoints
 * - can-i-deploy: Check if a schema change is safe for all consumers
 */

const { ContractValidator } = require('@cvt/cvt-sdk');
const { createMockAdapter } = require('@cvt/cvt-sdk/adapters');
const path = require('path');

const CVT_SERVER_ADDR = process.env.CVT_SERVER_ADDR || 'localhost:9550';
const SCHEMA_PATH = process.env.SCHEMA_PATH || path.join(__dirname, '../../producer/calculator-api.json');
const CONSUMER_ID = 'consumer-1';
const CONSUMER_VERSION = process.env.npm_package_version || '1.0.0';
const ENVIRONMENT = process.env.CVT_ENVIRONMENT || 'demo';

describe('Consumer Registration', () => {
  let validator;
  let mock;

  beforeAll(async () => {
    validator = new ContractValidator(CVT_SERVER_ADDR);
    await validator.registerSchema('calculator-api', SCHEMA_PATH);

    mock = createMockAdapter({
      validator,
      cache: true,
    });
  });

  afterAll(() => {
    if (mock) {
      mock.clearCache();
    }
    if (validator) {
      validator.close();
    }
  });

  beforeEach(() => {
    mock.clearInteractions();
  });

  describe('Auto-registration from interactions', () => {
    test('should capture interactions for auto-registration', async () => {
      await mock.fetch('http://calculator-api/add?a=5&b=3');
      await mock.fetch('http://calculator-api/subtract?a=10&b=4');

      const interactions = mock.getInteractions();
      expect(interactions.length).toBe(2);

      const opts = validator.buildConsumerFromInteractions(interactions, {
        consumerId: CONSUMER_ID,
        consumerVersion: CONSUMER_VERSION,
        environment: ENVIRONMENT,
        schemaVersion: '1.0.0',
        schemaId: 'calculator-api',
      });

      expect(opts.consumerId).toBe(CONSUMER_ID);
      expect(opts.usedEndpoints).toBeDefined();
      expect(opts.usedEndpoints.length).toBe(2);

      const endpoints = opts.usedEndpoints.map(e => `${e.method} ${e.path}`);
      expect(endpoints).toContain('GET /add');
      expect(endpoints).toContain('GET /subtract');
    });

    test('should register consumer from captured interactions', async () => {
      await mock.fetch('http://calculator-api/add?a=1&b=2');
      await mock.fetch('http://calculator-api/subtract?a=5&b=3');

      const interactions = mock.getInteractions();

      const consumer = await validator.registerConsumerFromInteractions(
        interactions,
        {
          consumerId: CONSUMER_ID,
          consumerVersion: CONSUMER_VERSION,
          environment: ENVIRONMENT,
          schemaVersion: '1.0.0',
          schemaId: 'calculator-api',
        }
      );

      expect(consumer).toBeDefined();
      expect(consumer.consumerId).toBe(CONSUMER_ID);
    });
  });

  describe('Manual registration', () => {
    test('should manually register consumer with explicit endpoints', async () => {
      const consumer = await validator.registerConsumer({
        consumerId: CONSUMER_ID,
        consumerVersion: CONSUMER_VERSION,
        schemaId: 'calculator-api',
        schemaVersion: '1.0.0',
        environment: ENVIRONMENT,
        usedEndpoints: [
          { method: 'GET', path: '/add', usedFields: ['result'] },
          { method: 'GET', path: '/subtract', usedFields: ['result'] },
        ],
      });

      expect(consumer).toBeDefined();
      expect(consumer.consumerId).toBe(CONSUMER_ID);
    });

    test('should list registered consumers', async () => {
      await validator.registerConsumer({
        consumerId: CONSUMER_ID,
        consumerVersion: CONSUMER_VERSION,
        schemaId: 'calculator-api',
        schemaVersion: '1.0.0',
        environment: ENVIRONMENT,
        usedEndpoints: [
          { method: 'GET', path: '/add', usedFields: ['result'] },
          { method: 'GET', path: '/subtract', usedFields: ['result'] },
        ],
      });

      const consumers = await validator.listConsumers('calculator-api', ENVIRONMENT);

      expect(consumers).toBeDefined();
      expect(Array.isArray(consumers)).toBe(true);
      expect(consumers.some(c => c.consumerId === CONSUMER_ID)).toBe(true);
    });
  });

  describe('Breaking change detection', () => {
    test('should check deployment safety with can-i-deploy', async () => {
      await validator.registerConsumer({
        consumerId: CONSUMER_ID,
        consumerVersion: CONSUMER_VERSION,
        schemaId: 'calculator-api',
        schemaVersion: '1.0.0',
        environment: ENVIRONMENT,
        usedEndpoints: [
          { method: 'GET', path: '/add', usedFields: ['result'] },
          { method: 'GET', path: '/subtract', usedFields: ['result'] },
        ],
      });

      const result = await validator.canIDeploy('calculator-api', '1.0.0', ENVIRONMENT);

      expect(result).toBeDefined();
      expect(result.safeToDeploy).toBeDefined();
    });
  });
});

File: consumer-3/src/test/java/demo/consumer3/RegistrationTest.java
package demo.consumer3;

import io.github.sahina.sdk.ContractValidator;
import io.github.sahina.sdk.ConsumerInfo;
import io.github.sahina.sdk.CanIDeployResult;
import io.github.sahina.sdk.adapters.CapturedInteraction;
import io.github.sahina.sdk.AutoRegisterConfig;
import io.github.sahina.sdk.AutoRegisterUtils;
import io.github.sahina.sdk.RegisterConsumerOptions;
import io.github.sahina.sdk.EndpointUsage;
import io.github.sahina.sdk.adapters.MockInterceptor;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * CONSUMER REGISTRATION
 * ---------------------
 * Registers endpoints used by consumer-3 for breaking change detection.
 *
 * Prerequisites:
 * - CVT server running (for schema registration and consumer operations)
 * - No producer service needed (uses MockInterceptor for interaction capture)
 *
 * When to use:
 * - Setting up breaking change detection
 * - Verifying can-i-deploy checks before deployments
 * - Documenting which endpoints a consumer depends on
 */
public class RegistrationTest {

    private static final String CVT_SERVER_ADDR =
            System.getenv().getOrDefault("CVT_SERVER_ADDR", "localhost:9550");
    private static final String SCHEMA_PATH =
            System.getenv().getOrDefault("SCHEMA_PATH", "../producer/calculator-api.json");
    private static final String SCHEMA_ID = "calculator-api";

    private static final String CONSUMER_ID = "consumer-3";
    private static final String CONSUMER_VERSION = "1.0.0";
    private static final String ENVIRONMENT =
            System.getenv().getOrDefault("CVT_ENVIRONMENT", "demo");

    private static ContractValidator validator;
    private static MockInterceptor mockInterceptor;
    private static OkHttpClient mockClient;

    @BeforeAll
    static void setUp() throws Exception {
        validator = new ContractValidator(CVT_SERVER_ADDR);
        validator.registerSchema(SCHEMA_ID, SCHEMA_PATH);

        mockInterceptor = new MockInterceptor(validator);
        mockClient = new OkHttpClient.Builder()
                .addInterceptor(mockInterceptor)
                .build();
    }

    @AfterAll
    static void tearDown() throws Exception {
        if (mockInterceptor != null) {
            mockInterceptor.clearCache();
        }
        if (validator != null) {
            validator.close();
        }
    }

    @BeforeEach
    void clearInteractions() {
        if (mockInterceptor != null) {
            mockInterceptor.clearInteractions();
        }
    }

    private void makeMockRequests() throws Exception {
        Request multiplyRequest = new Request.Builder()
                .url("http://calculator-api/multiply")
                .build();
        try (Response response = mockClient.newCall(multiplyRequest).execute()) {
            assertNotNull(response.body());
            response.body().string();
        }

        Request divideRequest = new Request.Builder()
                .url("http://calculator-api/divide")
                .build();
        try (Response response = mockClient.newCall(divideRequest).execute()) {
            assertNotNull(response.body());
            response.body().string();
        }
    }

    @Test
    void shouldCaptureInteractionsForAutoRegistration() throws Exception {
        makeMockRequests();

        List<CapturedInteraction> interactions = mockInterceptor.getInteractions();
        assertEquals(2, interactions.size(), "Expected 2 interactions for multiply and divide");

        AutoRegisterConfig config = AutoRegisterConfig.builder()
                .consumerId(CONSUMER_ID)
                .consumerVersion(CONSUMER_VERSION)
                .environment(ENVIRONMENT)
                .schemaId(SCHEMA_ID)
                .schemaVersion("1.0.0")
                .build();

        AutoRegisterUtils.BuildResult buildResult =
                validator.buildConsumerFromInteractions(interactions, config);

        assertNotNull(buildResult, "Build result should not be null");
        assertFalse(buildResult.hasError(),
                "Build should succeed without errors. Error: " + buildResult.getError());

        RegisterConsumerOptions options = buildResult.getOptions();
        assertNotNull(options, "Options should not be null");
        assertEquals(CONSUMER_ID, options.getConsumerId(),
                "Consumer ID should match");
        assertEquals(2, options.getUsedEndpoints().size(),
                "Should have 2 used endpoints (multiply and divide)");
    }

    @Test
    void shouldRegisterConsumerFromInteractions() throws Exception {
        makeMockRequests();

        List<CapturedInteraction> interactions = mockInterceptor.getInteractions();
        assertEquals(2, interactions.size(), "Expected 2 interactions");

        AutoRegisterConfig config = AutoRegisterConfig.builder()
                .consumerId(CONSUMER_ID)
                .consumerVersion(CONSUMER_VERSION)
                .environment(ENVIRONMENT)
                .schemaId(SCHEMA_ID)
                .schemaVersion("1.0.0")
                .build();

        ConsumerInfo consumer = validator.registerConsumerFromInteractions(interactions, config);

        assertNotNull(consumer, "Consumer info should not be null");
        assertEquals(CONSUMER_ID, consumer.getConsumerId(),
                "Registered consumer ID should match");
    }

    @Test
    void shouldRegisterConsumerWithExplicitEndpoints() throws Exception {
        RegisterConsumerOptions options = RegisterConsumerOptions.builder()
                .consumerId(CONSUMER_ID)
                .consumerVersion(CONSUMER_VERSION)
                .schemaId(SCHEMA_ID)
                .schemaVersion("1.0.0")
                .environment(ENVIRONMENT)
                .usedEndpoints(List.of(
                        new EndpointUsage("GET", "/multiply", List.of("result")),
                        new EndpointUsage("GET", "/divide", List.of("result"))
                ))
                .build();

        ConsumerInfo consumer = validator.registerConsumer(options);

        assertNotNull(consumer, "Consumer info should not be null");
        assertEquals(CONSUMER_ID, consumer.getConsumerId(),
                "Registered consumer ID should match");
    }

    @Test
    void shouldListRegisteredConsumers() throws Exception {
        // Register the consumer first
        RegisterConsumerOptions options = RegisterConsumerOptions.builder()
                .consumerId(CONSUMER_ID)
                .consumerVersion(CONSUMER_VERSION)
                .schemaId(SCHEMA_ID)
                .schemaVersion("1.0.0")
                .environment(ENVIRONMENT)
                .usedEndpoints(List.of(
                        new EndpointUsage("GET", "/multiply", List.of("result")),
                        new EndpointUsage("GET", "/divide", List.of("result"))
                ))
                .build();
        validator.registerConsumer(options);

        List<ConsumerInfo> consumers = validator.listConsumers(SCHEMA_ID, ENVIRONMENT);

        assertNotNull(consumers, "Consumer list should not be null");
        boolean found = consumers.stream()
                .anyMatch(c -> CONSUMER_ID.equals(c.getConsumerId()));
        assertTrue(found,
                "Registered consumer '" + CONSUMER_ID + "' should appear in consumer list");
    }

    @Test
    void shouldCheckDeploymentSafetyWithCanIDeploy() throws Exception {
        // Register the consumer first
        RegisterConsumerOptions options = RegisterConsumerOptions.builder()
                .consumerId(CONSUMER_ID)
                .consumerVersion(CONSUMER_VERSION)
                .schemaId(SCHEMA_ID)
                .schemaVersion("1.0.0")
                .environment(ENVIRONMENT)
                .usedEndpoints(List.of(
                        new EndpointUsage("GET", "/multiply", List.of("result")),
                        new EndpointUsage("GET", "/divide", List.of("result"))
                ))
                .build();
        validator.registerConsumer(options);

        CanIDeployResult result = validator.canIDeploy(SCHEMA_ID, "1.0.0", ENVIRONMENT);

        assertNotNull(result, "Can-i-deploy result should not be null");
    }
}
